### ISSUE TITLE
Replica failed

### ISSUE THREAD

**Issue by bachden** [CrunchyData/postgres-operator#2973]
My cluster stop syncing unexpectedly, and even I've trying to restart whole k8s cluster, it still persist, plz help.
below is error log from pg log in one replica pod:
![image](https://user-images.githubusercontent.com/1224571/148758666-5fc769fd-3f95-4d2b-bca3-a17f8dec533d.png)

UPDATE: (sorry when create this issue I were so hurry)
I'm using k8s with RKE, kubernetes version 1.19.15. This is my staging cluster, each node run on a vmWare VPS. 
Postgres-operator version 5.0.4, which has working normally for months, but yesterday above error occurred. 
Log files:
[[slave] postgresql-Mon.log](https://github.com/CrunchyData/postgres-operator/files/7842944/slave.postgresql-Mon.log)
[[master] postgresql-Mon.log](https://github.com/CrunchyData/postgres-operator/files/7842946/master.postgresql-Mon.log)
pod envs:
[slave-env.txt](https://github.com/CrunchyData/postgres-operator/files/7842978/slave-env.txt)
[master-env.txt](https://github.com/CrunchyData/postgres-operator/files/7842979/master-env.txt)

**Comment by jkatz** [comment]
There is insufficient information here to troubleshoot. Please see https://github.com/CrunchyData/postgres-operator/issues/new?template=bug_report.md for information that helps.

Additionally, I would also recommend ensuring that your cluster networking is stable.

**Comment by cbandy** [comment]
`$(patroni_name).pg01-pods` doesn't look right. `kubectl describe` the pod and look for `PATRONI_POSTGRESQL_CONNECT_ADDRESS` or `PATRONI_RESTAPI_CONNECT_ADDRESS`. These should both be defined with `$(PATRONI_NAME)` (uppercase).

https://github.com/CrunchyData/postgres-operator/blob/a7ce6ee7389967d02f7346e10e2f980b8a5cc21f/internal/patroni/config.go#L400-L401

**Comment by bachden** [comment]
> There is insufficient information here to troubleshoot. Please see https://github.com/CrunchyData/postgres-operator/issues/new?template=bug_report.md for information that helps.
> 
> Additionally, I would also recommend ensuring that your cluster networking is stable.

I've updated my post.

**Comment by bachden** [comment]
> `$(patroni_name).pg01-pods` doesn't look right. `kubectl describe` the pod and look for `PATRONI_POSTGRESQL_CONNECT_ADDRESS` or `PATRONI_RESTAPI_CONNECT_ADDRESS`. These should both be defined with `$(PATRONI_NAME)` (uppercase).
> 
> https://github.com/CrunchyData/postgres-operator/blob/a7ce6ee7389967d02f7346e10e2f980b8a5cc21f/internal/patroni/config.go#L400-L401

Thanks you, environment you can see from update in first post

**Comment by cbandy** [comment]
The [order of the environment variable definitions](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/) matters.

> Variables making use of others defined in the same context must come later in the list.

Do the two StatefulSets have their definitions in the same order?

```
kubectl describe sts/pg01-instance-t8zw | sed -n '/Environment:/,/Mounts:/p'
kubectl describe sts/pg01-instance-vct8 | sed -n '/Environment:/,/Mounts:/p'
```

The last thing to check is that the StatefulSet PodTemplate and Pod agree. A webhook (not PGO) may modify a Pod when the StatefulSet controller submits it.

```
kubectl describe sts/pg01-instance-t8zw | sed -n '/Environment:/,/Mounts:/p'
kubectl describe pod/pg01-instance-t8zw-0 | sed -n '/Environment:/,/Mounts:/p'
```

**Comment by bachden** [comment]
> The [order of the environment variable definitions](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/) matters.
> 
> > Variables making use of others defined in the same context must come later in the list.
> 
> Do the two StatefulSets have their definitions in the same order?
> 
> ```
> kubectl describe sts/pg01-instance-t8zw | sed -n '/Environment:/,/Mounts:/p'
> kubectl describe sts/pg01-instance-vct8 | sed -n '/Environment:/,/Mounts:/p'
> ```
> 
> The last thing to check is that the StatefulSet PodTemplate and Pod agree. A webhook (not PGO) may modify a Pod when the StatefulSet controller submits it.
> 
> ```
> kubectl describe sts/pg01-instance-t8zw | sed -n '/Environment:/,/Mounts:/p'
> kubectl describe pod/pg01-instance-t8zw-0 | sed -n '/Environment:/,/Mounts:/p'
> ```

I've delete the two replicas to make system work, so now only master is available for check. run following command: 
`kubectl --kubeconfig ~/.kube/config.live --context live-land01 describe sts/pg01-instance-t8zw | sed -n '/Environment:/,/Mounts:/p'`
-> output: 
`    Environment:
      PGDATA:  /pgdata/pg13
      PGHOST:  /tmp/postgres
      PGPORT:  5432
    Mounts:
    Environment:  <none>
    Mounts:
    Environment:
      LD_PRELOAD:                          /usr/lib64/libnss_wrapper.so
      NSS_WRAPPER_GROUP:                   /tmp/nss_wrapper/postgres/group
      NSS_WRAPPER_PASSWD:                  /tmp/nss_wrapper/postgres/passwd
      PATRONICTL_CONFIG_FILE:              /etc/patroni
      PATRONI_KUBERNETES_PORTS:            - name: postgres
                                             port: 5432
                                             protocol: TCP

      PATRONI_POSTGRESQL_CONFIG_DIR:       /pgdata/pg13
      PATRONI_POSTGRESQL_CONNECT_ADDRESS:  $(PATRONI_NAME).pg01-pods:5432
      PATRONI_POSTGRESQL_DATA_DIR:         /pgdata/pg13
      PATRONI_POSTGRESQL_LISTEN:           *:5432
      PATRONI_RESTAPI_CONNECT_ADDRESS:     $(PATRONI_NAME).pg01-pods:8008
      PATRONI_RESTAPI_LISTEN:              *:8008
      PGDATA:                              /pgdata/pg13
      PGHOST:                              /tmp/postgres
      PGPORT:                              5432
      PATRONI_NAME:                         (v1:metadata.name)
      PATRONI_KUBERNETES_POD_IP:            (v1:status.podIP)
    Mounts:
    Environment:  <none>
    Mounts:
    Environment:
      LD_PRELOAD:          /usr/lib64/libnss_wrapper.so
      NSS_WRAPPER_GROUP:   /tmp/nss_wrapper/postgres/group
      NSS_WRAPPER_PASSWD:  /tmp/nss_wrapper/postgres/passwd
    Mounts:
    Environment:
      CONFIG_DIR:                           /opt/cpm/conf
      EXPORTER_PG_DATABASE:                 postgres
      EXPORTER_PG_HOST:                     localhost
      EXPORTER_PG_PORT:                     5432
      EXPORTER_PG_USER:                     ccp_monitoring
      PGBACKREST_INFO_THROTTLE_MINUTES:     10
      PG_STAT_STATEMENTS_LIMIT:             20
      PG_STAT_STATEMENTS_THROTTLE_MINUTES:  -1
      POSTGRES_EXPORTER_PORT:               9187
      EXPORTER_PG_PASSWORD:                 <set to the key 'password' in secret 'pg01-monitoring'>  Optional: false
    Mounts:`

**Comment by cbandy** [comment]
This is definitely unusual. I don't think anything of ours rearranged those environment variables. To get back into a working state you can edit the affected StatefulSet and move the environment variable to the front of the array. The cluster should failover once then return to normal operation.

I believe this is a bug in Kubernetes server-side apply and will try to isolate it tomorrow.

**Comment by cbandy** [comment]
I was able to reproduce this on k8s 1.22 using similar steps as https://issue.k8s.io/104641. A fix for this was merged in 1.24 today and is being cherry-picked to 1.21, 1.22, and 1.23: https://pr.k8s.io/107565

```console
$ kubectl apply --server-side -f 00.yaml
deployment.apps/env-var-order serverside-applied
$ kubectl get deploy/env-var-order -o jsonpath='{.spec.template.spec.containers[*].env[*].name}{"\n"}'
original

$ kubectl apply --server-side -f 01.yaml
deployment.apps/env-var-order serverside-applied
$ kubectl get deploy/env-var-order -o jsonpath='{.spec.template.spec.containers[*].env[*].name}{"\n"}'
original first last
```

```diff
$ diff -u 00.yaml 01.yaml
--- 00.yaml     2022-01-14 23:07:35.087136641 +0000
+++ 01.yaml     2022-01-14 23:07:43.009135671 +0000
@@ -14,4 +14,6 @@
       - name: main
         image: busybox
         env:
+        - { name: first, value: any }
         - { name: original, value: any }
+        - { name: last, value: any }
```

**Comment by bachden** [comment]
> I was able to reproduce this on k8s 1.22 using similar steps as https://issue.k8s.io/104641. A fix for this was merged in 1.24 today and is being cherry-picked to 1.21, 1.22, and 1.23: https://pr.k8s.io/107565
> 
> ```
> $ kubectl apply --server-side -f 00.yaml
> deployment.apps/env-var-order serverside-applied
> $ kubectl get deploy/env-var-order -o jsonpath='{.spec.template.spec.containers[*].env[*].name}{"\n"}'
> original
> 
> $ kubectl apply --server-side -f 01.yaml
> deployment.apps/env-var-order serverside-applied
> $ kubectl get deploy/env-var-order -o jsonpath='{.spec.template.spec.containers[*].env[*].name}{"\n"}'
> original first last
> ```
> 
> ```diff
> $ diff -u 00.yaml 01.yaml
> --- 00.yaml     2022-01-14 23:07:35.087136641 +0000
> +++ 01.yaml     2022-01-14 23:07:43.009135671 +0000
> @@ -14,4 +14,6 @@
>        - name: main
>          image: busybox
>          env:
> +        - { name: first, value: any }
>          - { name: original, value: any }
> +        - { name: last, value: any }
> ```

Thanks a lot,
